pennylane/estimator/wires_manager.py

```diff
@@ -0,0 +1,201 @@
+        tight_budget (bool): Determines whether extra clean qubits can be allocated when they
+            exceed the available amount. The default is ``False``.
+
+    **Example**
+
+    >>> q = _WireResourceManager(
+    ...             work_wires={"clean": 2, "dirty": 2},
+    ...             tight_budget=False,
+    ...     )
+    >>> print(q)
+    _WireResourceManager(clean=2, dirty=2, logic=0, tight_budget=False)
+
+    """
+
+    def __init__(self, work_wires: int | dict, algo_wires=0, tight_budget=False) -> None:
+
+        if isinstance(work_wires, dict):
+            clean_wires = work_wires["clean"]
+            dirty_wires = work_wires["dirty"]
+        else:
+            clean_wires = work_wires
+            dirty_wires = 0
+
+        self.tight_budget = tight_budget
+        self._logic_qubit_counts = algo_wires
+        self._clean_qubit_counts = clean_wires
+        self._dirty_qubit_counts = dirty_wires
+
+    def __str__(self):
+        return (
+            f"_WireResourceManager(clean qubits={self._clean_qubit_counts}, dirty qubits={self._dirty_qubit_counts}, "
+            f"algorithmic qubits={self._logic_qubit_counts}, tight budget={self.tight_budget})"
+        )
+
+    def __repr__(self) -> str:
+        work_wires_str = repr(
+            {"clean": self._clean_qubit_counts, "dirty": self._dirty_qubit_counts}
+        )
+        return (
+            f"_WireResourceManager(work_wires={work_wires_str}, algo_wires={self._logic_qubit_counts}, "
+            f"tight_budget={self.tight_budget})"
+        )
+
+    def __eq__(self, other: object) -> bool:
+        return (
+            isinstance(other, self.__class__)
+            and (self.clean_qubits == other.clean_qubits)
+            and (self.dirty_qubits == other.dirty_qubits)
+            and (self.algo_qubits == other.algo_qubits)
+            and (self.tight_budget == other.tight_budget)
+        )
+
+    @property
+    def clean_qubits(self):
+        r"""Returns the number of clean qubits."""
+        return self._clean_qubit_counts
+
+    @property
+    def dirty_qubits(self):
+        r"""Returns the number of dirty qubits."""
+        return self._dirty_qubit_counts
+
+    @property
+    def algo_qubits(self):
+        r"""Returns the number of algorithmic qubits."""
+        return self._logic_qubit_counts
+
+    @property
+    def total_qubits(self):
+        r"""Returns the number of total qubits."""
+        return self._clean_qubit_counts + self._dirty_qubit_counts + self.algo_qubits
+
+    @algo_qubits.setter
+    def algo_qubits(self, count: int):  # these get set manually, the rest are dynamically updated
+        r"""Setter for algorithmic qubits."""
+        self._logic_qubit_counts = count
+
+    def allocate_qubits(self, num_qubits: int):
+        r"""Allocates extra clean qubits.
+
+        Args:
+            num_qubits(int): number of qubits to be allocated
+
+        """
+        self._clean_qubit_counts += num_qubits
```

Comment:
Wait, I'm confused, doesn't `self.clean_qubits` refer to the number of clean qubits still available? If you allocate qubits, shouldn't that reduce the number of qubits available?